### PR TITLE
MEN-3489, MEN-3499: add section about software versioning

### PR DESCRIPTION
**Pull request by tranchitella**

### CODE REVIEW

File: 06.Artifact-creation/09.Software-versioning/docs.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    --software-name myapp \
+    --software-version 1.0.0 \
+   ...
+```
+The command above will generate the following versioning key/value pair:
+
+<!--AUTOVERSION: "myfs.myapp.version=%"/ignore-->
+```
+myfs.myapp.version=1.0.0
+```
```
Comment (kacf) [inline 06.Artifact-creation/09.Software-versioning/docs.md:129]: I think that we will also need to cover `clears_artifact_provides` here. These definitely belong to advanced usage, but they are an important part of the feature set. I suggest a new sub page with an introduction marking this as an advanced section, and then a few example use cases. Some of the use cases I can think of:

* Alternative rootfs-image update methods, based on Update Module:
  ```bash
    mender-artifact write module-image -T custom-root-image --software-filesystem rootfs-image --no-default-clears-provides --clears-provides 'rootfs-image.*'
    ```
* Updating other partitions:
  ```bash
  mender-artifact write module-image -T part2-image --software-filesystem part2-image --no-default-clears-provides --clears-provides 'part2-image.*'
  ```
* Proxy updates:
  * You have multiple connected devices that you can update both fully and partially. In this scenario, for the full updates to report correctly, they need to be created like this:
    ```bash
    mender-artifact write module-image -T full-proxy-update --software-filesystem device1-fs --no-default-clears-provides --clears-provides 'device1-fs.*'
    ```

**Review by kacf — CHANGES_REQUESTED** [review]
Well written!

File: 06.Artifact-creation/09.Software-versioning/docs.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+taxonomy:
+    category: docs
+    label: tutorial
+---
+
+## Introduction
+
+Reporting the current software version running on a device can be challenging when mixing full system updates and application updates. On top of this, installing software on different partitions than the root filesystem requires additional flexibility in storing and reporting the current version of software installed into the device.
+
+For these reasons, the Mender Artifact tool supports several options to customize the current software version's reporting, both for full system updates and Update modules.
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply 06.Artifact-creation/09.Software-versioning/docs.md:12]: Ack

File: 06.Artifact-creation/09.Software-versioning/docs.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+The software versioning is stored in the Mender artifacts as *Provides*, and is automatically reported by the Mender Client as inventory data. See below for further details on how to override the default software versioning keys when generating the artifacts.
+
+## Full system updates
+
+When creating a full system update, the Mender Artifact tool automatically generates a *Provides* entry using the key `rootfs-image.version` and the artifact name as value. By default, the Mender Client will automatically report such a key/value pair as inventory data.
+
+It is possible to customize this behaviour overriding the base key (`rootfs-image`) with a custom one using the `--software-filesystem` flag of `mender-artifact`. It is also possible to set a specific name as a suffix for the versioning key, using the `--software-name` flag. If provided, the generated software version key will be `rootfs-image.NAME.version`. Additionally, to override the value of the versioning key, which defaults to the artifact's name, you can use the `--software-version` flag.
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply 06.Artifact-creation/09.Software-versioning/docs.md:20]: Ack

File: 06.Artifact-creation/09.Software-versioning/docs.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+   ...
+```
+The command above will generate the following versioning ekey/value pair:
+
+<!--AUTOVERSION: "rootfs-image.script.version=%"/ignore-->
+```
+rootfs-image.script.version=1.0.0
+```
+
+To override the software name in the versioning key, use the `--software-name" flag as follows:
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply 06.Artifact-creation/09.Software-versioning/docs.md:96]: Ack

File: 06.Artifact-creation/09.Software-versioning/docs.md (hunk 1 above)
Comment (tranchitella) [reply 06.Artifact-creation/09.Software-versioning/docs.md:129]: done

File: 06.Artifact-creation/09.Software-versioning/docs.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kacf) [reply 06.Artifact-creation/09.Software-versioning/docs.md:12]: You changed the header, but you didn't change it here.

File: 06.Artifact-creation/09.Software-versioning/docs.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+<!--AUTOVERSION: "myfs.myapp.version=%"/ignore-->
+```
+myfs.myapp.version=1.0.0
+```
+
+## Advanced usage
+
+To support advanced software reporting needs, the Mender Artifact tool also supports the so-named *Clear provides* payload field, a list of patterns that remove the *Provides* entries matching them.
+
+Each pattern can contain multiple placeholders, each of them expressed using asterisks (`*`), which match a sequence of zero or more characters. For example, the pattern `rootfs-image.*` matches `rootfs-image.version` and `rootfs-image.checksum` but not `datafs-image.checksum` nr `rootfs-image`.
```
Comment (kacf) [inline 06.Artifact-creation/09.Software-versioning/docs.md:100]: "...but not `datafs-image.checksum` **nor** `rootfs-image`."

File: 06.Artifact-creation/09.Software-versioning/docs.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+Mender Artifacts supports the following options:
+
+* `--clears-provides`: adds the specified pattern to the *Clear provides*
+
+* `--delete-clears-provides`: removes the specified pattern from the *Clear provides*
```
Comment (kacf) [inline 06.Artifact-creation/09.Software-versioning/docs.md:106]: This is for `mender-artifact modify` only, and doesn't work in `write` mode. I think it's ok not to describe this one, and let people discover it when they use `modify`.

File: 06.Artifact-creation/09.Software-versioning/docs.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+* `--no-default-clears-provides`: disables the automatic generation of *Clear provides* to gain full control on the patterns.
+
+For example, to provide an alternative rootfs-image update method based on Update Module, you would need an artifact that clears the `rootfs-image.version` and `rootfs-image.checksum` *Provides* keys.
```
Comment (kacf) [inline 06.Artifact-creation/09.Software-versioning/docs.md:110]: Minor: "to provide an alternative rootfs-image update method based on **an** Update Module"

Comment (kacf) [inline 06.Artifact-creation/09.Software-versioning/docs.md:110]: Our style guide says that "artifact" should be capitalized "Artifact" in all places where it makes sense. This should probably be changed throughout.

**Review by kacf — CHANGES_REQUESTED** [review]
Also very well written! All my comments are minor things, the next one will be "the one", I think!

**Comment by tranchitella** [comment]
@kacf ready for another round, don't be shy if you find something else, I'm okay with multiple rounds of the review! :)

**Comment by kacf** [comment]
@tranchitella: Uhm, did you actually push? Nothing changed AFAICS.

**Comment by tranchitella** [comment]
@kacf sorry, push failed because of network issues here and I did not notice it

**Review by kacf — APPROVED** [review]
:+1:

File: 06.Artifact-creation/09.Software-versioning/docs.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+For example, to explicitly set the software version for an application update to the value "1.0.0", you can use the following Mender Artifact command line:
+
+<!--AUTOVERSION: "software-version %"/ignore-->
+```bash
+$ mender-artifact write module-image \
+    -T script
+    --software-version 1.0.0 \
+   ...
+```
+The command above will generate the following versioning ekey/value pair:
```
Comment (lluiscampos) [inline 06.Artifact-creation/09.Software-versioning/docs.md:54]: Typo: `key/value`

File: 06.Artifact-creation/09.Software-versioning/docs.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+Each pattern can contain multiple placeholders, each of them expressed using asterisks (`*`), which match a sequence of zero or more characters. For example, the pattern `rootfs-image.*` matches `rootfs-image.version` and `rootfs-image.checksum` but not `datafs-image.checksum` nor `rootfs-image`.
+
+Mender Artifacts supports the following options:
+
+* `--clears-provides`: adds the specified pattern to the *Clear provides*
+
+* `--no-default-clears-provides`: disables the automatic generation of *Clear provides* to gain full control on the patterns.
+
+For example, to provide an alternative rootfs-image update method based on an Update Module, you would need an Artifact that clears the `rootfs-image.version` and `rootfs-image.checksum` *Provides* keys.
```
Comment (lluiscampos) [inline 06.Artifact-creation/09.Software-versioning/docs.md:108]: Inconsistency: `Update module` here or `Update Module` elsewhere.

**Review by lluiscampos — COMMENTED** [review]
Indeed, very well written. Personally, I understood quite well the feature now (from the user point of view).

Some comments below, but as arrived very late to the party just follow-up if you have the capacity.

**Comment by tranchitella** [comment]
@lluiscampos follow-up PR: https://github.com/mendersoftware/mender-docs/pull/1201

**Pull request merged**